### DISCUSSION TITLE
Rule - partial input.string()

### DISCUSSION THREAD

**Discussion by theduke** [taocpp/PEGTL#25] [Q&A] [answered]
Given a rule like this:
`
struct module_decl : seq< string< 'm', 'o', 'd', 'u', 'l', 'e' >, one< ' ' >, identifier > {};
`

Is there an easy way to have string<...> discarded from the input from input.string()?
Something like:

`
struct module_decl : seq< skip< string< 'm', 'o', 'd', 'u', 'l', 'e' >, one< ' ' > >, identifier > {};
...
action:
  // input: "module my_identifier"
  input.string() == "my_identifier"
`

This would make building up an AST quite a bit easier.

**Comment by d-frey** [comment, marked as answer]
The usual trick is to apply the action only to the part you are interested in:

```
struct module_identifier : identifier {};
struct module_decl : seq< string< 'm', 'o', 'd', 'u', 'l', 'e', ' ' >, module_identifier > {};
```

and then specialize your action for `module_identifier`. The additional `struct` is used to make sure the action is only applied for this case, not for other uses of `identifier`.

Does this work for you?

**Comment by theduke** [comment]
Good trick, I could have thought of that myself...
Thanks. That makes it a bit easier.

ps: you have any plans for making building an AST more convenient, similar to the return values offered by bison?
It would be nice if I could return something (eg a pointer to some AST node) inside an action, and then pick it up in a 'parent' rules action with `input.data(3)` or something similar.

pps: great project. It's definitely more convenient to use PEGTL than to either wire up flex and bison with C++ and build management or to pull in the beast that is boost::spirit.

**Comment by d-frey** [comment]
Glad it helped :)

Re: ps: In fact we are planning on adding a generic AST builder to the PEGTL in the future, together with some other enhancements. I'm not sure which specific form it will take, so I can not compare it to Bison or anything else yet. It might take some time though, as we are currently busy finishing some new project (https://github.com/taocpp/json) which is using the PEGTL to parse JSON. And there are other things in the pipeline as well and, as usual, not enough time.

Re: pps: Thank you!
